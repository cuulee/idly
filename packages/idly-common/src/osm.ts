export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
