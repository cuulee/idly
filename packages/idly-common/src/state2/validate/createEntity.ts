import { Either } from 'monet';
import { nodeFactory } from '../../osm/entityFactory/nodeFactory';
import { relationFactory } from '../../osm/entityFactory/relationFactory';
import { wayFactory } from '../../osm/entityFactory/wayFactory';
import {
  Attributes,
  Entity,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Relation,
  RelationMember,
  Tags,
  Way,
} from '../../osm/structures';
import { Err, ErrorWhen } from '../helper';
import {
  checkValidNode,
  checkValidRelation,
  checkValidWay,
} from '../validation';

export interface WayLike {
  readonly id: EntityId;
  readonly type?: 'way';
  readonly tags?: Tags;
  readonly attributes?: Attributes;
  readonly nodes?: EntityId[] | ReadonlyArray<EntityId>;
}
export interface RelLike {
  readonly id: EntityId;
  readonly type?: 'relation';
  readonly tags?: Tags;
  readonly members?: RelationMember[] | ReadonlyArray<RelationMember>;
  readonly attributes?: Attributes;
}

export interface NodeLike {
  readonly id: EntityId;
  readonly type?: 'node';
  readonly tags?: Tags;
  readonly loc?: LngLat;
  readonly attributes?: Attributes;
}
export type EntityLike = NodeLike | WayLike | RelLike;

export const parseIfString = (r: EntityLike | string) =>
  typeof r === 'string' ? (JSON.parse(r) as EntityLike) : r;

export function createEntity(
  entityLike: EntityLike | string,
  type?: EntityType
): Entity {
  return Err.right(entityLike)
    .map(parseIfString)
    .flatMap(
      ErrorWhen(
        r => !!type && type !== r.type,
        new Error('Entity must be ' + type)
      )
    )
    .flatMap(
      ErrorWhen(r => !r.id, new Error('Entity does not have a valid id'))
    )
    .flatMap(entityFrom)
    .cata(err => {
      throw err;
    }, r => r);
}

function entityFrom(entity: any): Either<Error, Entity> {
  switch (entity.type) {
    case EntityType.NODE: {
      return checkValidNode(entity)
        ? Err.right(nodeFactory(entity))
        : Err.left(new Error('Invalid node'));
    }

    case EntityType.WAY: {
      return checkValidWay(entity)
        ? Err.right(wayFactory(entity))
        : Err.left(new Error('Invalid Way'));
    }

    case EntityType.RELATION: {
      return checkValidRelation(entity)
        ? Err.right(relationFactory(entity))
        : Err.left(new Error('Invalid Relation'));
    }
    default:
      return Err.left(new Error('Entity type is unknown'));
  }
}

export const createNode = (entityLike: NodeLike): Node =>
  createEntity({ ...entityLike }, EntityType.NODE) as Node;

export const createWay = (entityLike: WayLike): Way =>
  createEntity({ ...entityLike }, EntityType.WAY) as Way;

export const createRelation = (entityLike: RelLike): Relation =>
  createEntity({ ...entityLike }, EntityType.RELATION) as Relation;
