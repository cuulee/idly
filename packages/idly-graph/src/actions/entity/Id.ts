export class Id {
  private readonly id: number;
  constructor(id: number) {
    /* tslint:disable */
    this.id = id;
    /* tslint:enable */
  }
  public getId(): number {
    return this.id;
  }
  public next(): Id {
    return new Id(this.id + 1);
  }
}
