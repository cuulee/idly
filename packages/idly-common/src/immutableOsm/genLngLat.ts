import { recordify } from "typed-immutable-record";
import { $LngLat } from "../immutableOsm/immutableOsm";
import { LngLat } from "../osm/structures";

export function $genLngLat(l: LngLat): $LngLat {
  return recordify<LngLat, $LngLat>(l);
}
