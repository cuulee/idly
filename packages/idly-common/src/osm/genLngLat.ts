import { LngLat } from "../osm/structures";

export function genLngLat(
  obj: { lon: number; lat: number } | [number, number],
): LngLat {
  if (Array.isArray(obj)) {
    return { lon: obj[0], lat: obj[1] };
  }
  return { ...obj };
}
