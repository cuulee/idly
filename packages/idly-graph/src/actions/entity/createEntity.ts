import { attributesGen } from 'idly-common/lib/osm/attributesGen';
import { entityFromString } from 'idly-common/lib/osm/entityFromString';
import { entityToString } from 'idly-common/lib/osm/entityToString';
import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Relation,
  RelationMember,
  Tags,
  Way,
} from 'idly-common/lib/osm/structures';

export interface NodeLike {
  readonly id: EntityId;
  readonly type?: 'node';
  readonly tags?: Tags;
  readonly loc?: LngLat;
  readonly attributes?: Attributes;
}
export interface WayLike {
  readonly id: EntityId;
  readonly type?: 'way';
  readonly tags?: Tags;
  readonly attributes?: Attributes;
  readonly nodes?: EntityId[];
}
export interface RelLike {
  readonly id: EntityId;
  readonly type?: 'relation';
  readonly tags?: Tags;
  readonly members?: RelationMember[];
  readonly attributes?: Attributes;
}

export function createEntity(
  entityLike: NodeLike | WayLike | RelLike | string,
) {
  if (typeof entityLike === 'string') {
    entityLike = JSON.parse(entityLike) as NodeLike | WayLike | RelLike;
  }

  if (!entityLike.id || !entityLike.type) {
    throw new Error('not a valid entity');
  }

  return entityFromString(entityToString(entityLike));
}

export function createNode(entityLike: NodeLike): Node {
  return createEntity({ ...entityLike, type: 'node' }) as Node;
}

export function createWay(entityLike: WayLike): Way {
  return createEntity({ ...entityLike, type: 'way' }) as Way;
}

export function createRelation(entityLike: RelLike): Relation {
  return createEntity({ ...entityLike, type: 'relation' }) as Relation;
}
