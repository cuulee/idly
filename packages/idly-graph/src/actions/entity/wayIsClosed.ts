import { Way } from 'idly-common/lib/osm/structures';

export function wayIsClosed(way: Way): boolean {
  const first = way.nodes[0];
  const last = way.nodes[way.nodes.length - 1];
  return way.nodes.length > 1 && first === last;
}
