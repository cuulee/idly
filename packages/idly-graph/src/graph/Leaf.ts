import { weakCache } from 'idly-common/lib/misc/weakCache';

import {
  EntityType,
  Node,
  Relation,
  Way,
} from 'idly-common/lib/osm/structures';
import { Entity } from 'idly-common/lib/osm/structures';

import {
  createEntity,
  NodeLike,
  RelLike,
  WayLike,
} from '../actions/entity/createEntity';

let LeafCache = new WeakMap<Entity, Leaf>();

const mStrongCache = new Map();

export type EntityKeys = keyof (Node & Way & Relation);

export class Leaf {
  public static create(entity: Entity, ancestor?: Leaf): Leaf {
    const inCache = LeafCache.get(entity);

    if (inCache) {
      return inCache;
    }

    const toRet = new Leaf(entity, ancestor);
    if (LeafCache) {
      // tslint:disable-next-line:no-expression-statement
      LeafCache.set(entity, toRet);
    }
    return toRet;
  }

  public static fromString(leafString: string): Leaf | undefined {
    const entities: Entity[] = JSON.parse(leafString)
      .map(createEntity)
      .reverse();

    let prevLeaf;
    for (const entity of entities) {
      // tslint:disable-next-line:no-expression-statement
      prevLeaf = Leaf.create(entity, prevLeaf);
    }
    return prevLeaf;
  }

  private static _create(entity: Entity, ancestor: Leaf | undefined): Leaf {
    return new Leaf(entity, ancestor);
  }

  private readonly entity: Entity;

  private readonly ancestor: Leaf | undefined;

  constructor(
    entity: Entity,
    ancestor: Leaf | undefined,
    cache?: WeakMap<any, any>,
  ) {
    /* tslint:disable */
    this.ancestor = ancestor;
    this.entity = entity;
    /* tslint:enable */
  }

  public getAncestor(): Leaf | undefined {
    return this.ancestor;
  }

  public getEntity(): Entity {
    return this.entity;
  }
  /**
   * accepts an EntityLike
   */
  public map(cb: (e: Entity) => any): Leaf {
    const partEntity = cb(this.entity);

    if (partEntity.id && partEntity.id !== this.entity.id) {
      throw new Error('cannot modify entity id');
    }

    const newEntity = createEntity(partEntity);

    if (newEntity !== this.entity) {
      return Leaf.create(newEntity, this);
    }
    return this;
  }

  /** derived methods */
  public toString(): string {
    return leafToString(this);
  }

  public getAllAncestors(): Leaf[] {
    return getAllAncestors(this);
  }

  public getFamily(): Leaf[] {
    return getFamily(this);
  }

  public getDependencies(): string[] {
    return getDependencies(this);
  }

  public getFirstAncestor(): Leaf | undefined {
    return getFirstAncestor(this);
  }

  public isFirstGeneration(): boolean {
    return isFirstGeneration(this);
  }
}

/**
 * only returns ancestors i.e. excluding self
 * @param leaf
 */
export let getAllAncestors = weakCache((leaf: Leaf): Leaf[] => {
  const ancestor = leaf.getAncestor();
  if (!ancestor) {
    return [];
  }
  return getFamily(ancestor);
});

/**
 * returns self + ancestors
 */
export let getFamily = weakCache((leaf: Leaf): Leaf[] => {
  const toReturn = [leaf];
  const ancestor = leaf.getAncestor();
  if (!ancestor) {
    return toReturn;
  }
  return toReturn.concat(getFamily(ancestor));
});

export let getAllEntities = (leaf: Leaf): Entity[] =>
  getFamily(leaf).map(l => l.getEntity());

export let getDependencies = (leaf: Leaf): string[] => {
  const dep = new Set();
  const entity = leaf.getEntity();
  switch (entity.type) {
    /* tslint:disable */
    case EntityType.NODE: {
      dep.add(entity.id);
      break;
    }
    case EntityType.WAY: {
      entity.nodes.forEach(n => dep.add(n));
      break;
    }
    case EntityType.RELATION: {
      entity.members.forEach(m => {
        if (m.id) {
          dep.add(m.id);
        }
      });
      break;
    }
    /* tslint:enable */
  }
  return [...dep];
};

export let getFirstAncestor = (leaf: Leaf): Leaf | undefined => {
  const ancestors = getAllAncestors(leaf);
  return ancestors[ancestors.length - 1];
};

export let isFirstGeneration = (leaf: Leaf) =>
  leaf.getAncestor() ? false : true;

/**
 * @notes need to stringify entities twice, to use caching to have just one
 * instance of entity running.
 * @param leaf
 */
export let leafToString = (leaf: Leaf): string =>
  JSON.stringify(getAllEntities(leaf).map(r => JSON.stringify(r)));

/** cache */
export function cacheAccess(): {
  readonly LeafCache: WeakMap<any, any>;
} {
  return { LeafCache };
}

export function setCache(cache: {
  readonly mWeakCache: WeakMap<any, any>;
  readonly mStrongCache: Map<any, any>;
}): void {
  // tslint:disable-next-line:no-expression-statement
  LeafCache = cache.mWeakCache;
}
