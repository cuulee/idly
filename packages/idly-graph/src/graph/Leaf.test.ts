import { genLngLat } from 'idly-common/lib/osm/genLngLat';

import { createNode, createWay } from '../actions/entity/createEntity';

import { Leaf } from './Leaf';

describe('map', () => {
  test('maintains the same instance', () => {
    const e = createNode({ id: 'n1' });
    const l1 = Leaf.create(e);
    const l2 = l1.map(entity => entity);
    expect(l1).toBe(l2);

    const l3 = Leaf.create(e);
    const l4 = l3.map(entity => entity);
    expect(l3).toBe(l4);
    expect(l3).toBe(l1);
  });

  test('different instance', () => {
    const e = createNode({ id: 'n1' });
    const l1 = Leaf.create(e);
    const l2 = l1.map(entity => {
      return createNode({ id: 'n1', tags: { K: 'K' } });
    });
    expect(l2.getAncestor()).toBe(l1);
    expect(l1.getEntity()).not.toEqual(l2.getEntity());
  });
  test('divergence of leaf', () => {
    const e = createNode({ id: 'n1' });
    const l1 = Leaf.create(e);
    const l2 = l1.map(entity => {
      return createNode({ id: 'n1', tags: { K: 'K' } });
    });
    const l2d = l1.map(entity => {
      return createNode({ id: 'n1', tags: { K: 'Kj' } });
    });
  });

  test('old ancestor become new ancestor', () => {
    // this test is important as it tells the ancestors
    // can be reduced, ancestors are minimum steps needed to
    // make the entity
    const n1 = createNode({ id: 'n1' });
    const n2 = createNode({ id: 'n1', tags: { K: '2' } });
    const n3 = createNode({ id: 'n1', tags: { K: '3' } });
    const l1 = Leaf.create(n1);

    const l2 = l1.map(entity => {
      return n2;
    });
    const l3 = l2.map(entity => {
      return n3;
    });
    expect(l3.map(entity => n1)).toBe(l1);
  });

  test('be able to return partial entity', () => {});
});

describe('ancestors', () => {
  const e = createNode({ id: 'n1' });
  const l1 = Leaf.create(e);

  const l2 = l1.map(entity => {
    return createNode({ id: 'n1', tags: { K: 'K' } });
  });

  const l3 = l2.map(entity => {
    return createNode({ id: 'n1', tags: { K: 'K', T: 'T' } });
  });

  test('getAncestor', () => {
    expect(l1.getAncestor()).toBeUndefined();
    expect(l2.getAncestor()).toBe(l1);
    expect(l3.getAncestor()).toBe(l2);
  });

  test('getAllAncestors', () => {
    expect(l2.getAllAncestors()).toEqual([l1]);
    expect(l3.getAllAncestors()).toEqual([l2, l1]);
  });
});

describe('dependencies', () => {
  const e = createWay({ id: 'n1', nodes: ['n1', 'n2'] });
  const l1 = Leaf.create(e);
  test('dep', () => {
    expect(l1.getDependencies()).toEqual(e.nodes);
  });
});

describe('isFirstGeneration', () => {
  const e = createWay({ type: 'way', id: 'n1', nodes: ['n1', 'n2'] });
  const l1 = Leaf.create(e);
  test('is first', () => {
    expect(l1.isFirstGeneration()).toEqual(true);
  });
});

describe('stringifying', () => {
  const e = createNode({ id: 'n1' });
  const l1 = Leaf.create(e);
  const l2 = l1.map(entity => entity);
  test('simple stringify', () => {
    expect(l1.toString()).toMatchSnapshot();
    expect(JSON.parse(l1.toString())).toMatchSnapshot();
    expect(JSON.parse(l1.toString()).map(JSON.parse)).toMatchSnapshot();
  });

  const leaf1 = Leaf.create(createNode({ id: 'n-a', tags: { k: 'n-a1' } }));
  const leaf2 = leaf1.map(entity => {
    return createNode({ id: 'n-a', tags: { k: 'n-a2' } });
  });

  const leaf3 = leaf2.map(entity => {
    return createNode({ id: 'n-a', tags: { k: 'n-a3' } });
  });
  const leaf4 = leaf3.map(entity => {
    return createNode({
      id: 'n-a',
      loc: genLngLat([1, 2]),
      tags: { k: 'n-a4' },
      type: 'node',
    });
  });
  test('3 parents', () => {
    expect(leaf4.toString()).toMatchSnapshot();
    expect(JSON.parse(leaf4.toString())).toMatchSnapshot();
    expect(JSON.parse(leaf4.toString()).map(JSON.parse)).toMatchSnapshot();

    const newLeaf4 = Leaf.fromString(leaf4.toString()) as Leaft;
    expect(newLeaf4).toEqual(leaf4);
    expect(newLeaf4).toBe(leaf4);

    expect(newLeaf4.getAncestor()).toEqual(leaf3);
    expect((newLeaf4.getAncestor() as Leaf).getAncestor()).toEqual(leaf2);
    expect(
      ((newLeaf4.getAncestor() as Leaf).getAncestor() as Leaf).getAncestor(),
    ).toEqual(leaf1);
  });
});
