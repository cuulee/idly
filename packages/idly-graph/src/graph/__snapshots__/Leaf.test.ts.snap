// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifying 3 parents 1`] = `"[\\"{\\\\\\"attributes\\\\\\":{},\\\\\\"id\\\\\\":\\\\\\"n-a\\\\\\",\\\\\\"loc\\\\\\":{\\\\\\"lon\\\\\\":1,\\\\\\"lat\\\\\\":2},\\\\\\"tags\\\\\\":{\\\\\\"k\\\\\\":\\\\\\"n-a4\\\\\\"},\\\\\\"type\\\\\\":\\\\\\"node\\\\\\"}\\",\\"{\\\\\\"attributes\\\\\\":{},\\\\\\"id\\\\\\":\\\\\\"n-a\\\\\\",\\\\\\"loc\\\\\\":{\\\\\\"lon\\\\\\":0,\\\\\\"lat\\\\\\":0},\\\\\\"tags\\\\\\":{\\\\\\"k\\\\\\":\\\\\\"n-a3\\\\\\"},\\\\\\"type\\\\\\":\\\\\\"node\\\\\\"}\\",\\"{\\\\\\"attributes\\\\\\":{},\\\\\\"id\\\\\\":\\\\\\"n-a\\\\\\",\\\\\\"loc\\\\\\":{\\\\\\"lon\\\\\\":0,\\\\\\"lat\\\\\\":0},\\\\\\"tags\\\\\\":{\\\\\\"k\\\\\\":\\\\\\"n-a2\\\\\\"},\\\\\\"type\\\\\\":\\\\\\"node\\\\\\"}\\",\\"{\\\\\\"attributes\\\\\\":{},\\\\\\"id\\\\\\":\\\\\\"n-a\\\\\\",\\\\\\"loc\\\\\\":{\\\\\\"lon\\\\\\":0,\\\\\\"lat\\\\\\":0},\\\\\\"tags\\\\\\":{\\\\\\"k\\\\\\":\\\\\\"n-a1\\\\\\"},\\\\\\"type\\\\\\":\\\\\\"node\\\\\\"}\\"]"`;

exports[`stringifying 3 parents 2`] = `
Array [
  "{\\"attributes\\":{},\\"id\\":\\"n-a\\",\\"loc\\":{\\"lon\\":1,\\"lat\\":2},\\"tags\\":{\\"k\\":\\"n-a4\\"},\\"type\\":\\"node\\"}",
  "{\\"attributes\\":{},\\"id\\":\\"n-a\\",\\"loc\\":{\\"lon\\":0,\\"lat\\":0},\\"tags\\":{\\"k\\":\\"n-a3\\"},\\"type\\":\\"node\\"}",
  "{\\"attributes\\":{},\\"id\\":\\"n-a\\",\\"loc\\":{\\"lon\\":0,\\"lat\\":0},\\"tags\\":{\\"k\\":\\"n-a2\\"},\\"type\\":\\"node\\"}",
  "{\\"attributes\\":{},\\"id\\":\\"n-a\\",\\"loc\\":{\\"lon\\":0,\\"lat\\":0},\\"tags\\":{\\"k\\":\\"n-a1\\"},\\"type\\":\\"node\\"}",
]
`;

exports[`stringifying 3 parents 3`] = `
Array [
  Object {
    "attributes": Object {},
    "id": "n-a",
    "loc": Object {
      "lat": 2,
      "lon": 1,
    },
    "tags": Object {
      "k": "n-a4",
    },
    "type": "node",
  },
  Object {
    "attributes": Object {},
    "id": "n-a",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {
      "k": "n-a3",
    },
    "type": "node",
  },
  Object {
    "attributes": Object {},
    "id": "n-a",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {
      "k": "n-a2",
    },
    "type": "node",
  },
  Object {
    "attributes": Object {},
    "id": "n-a",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {
      "k": "n-a1",
    },
    "type": "node",
  },
]
`;

exports[`stringifying simple stringify 1`] = `"[\\"{\\\\\\"attributes\\\\\\":{},\\\\\\"id\\\\\\":\\\\\\"n1\\\\\\",\\\\\\"loc\\\\\\":{\\\\\\"lon\\\\\\":0,\\\\\\"lat\\\\\\":0},\\\\\\"tags\\\\\\":{},\\\\\\"type\\\\\\":\\\\\\"node\\\\\\"}\\"]"`;

exports[`stringifying simple stringify 2`] = `
Array [
  "{\\"attributes\\":{},\\"id\\":\\"n1\\",\\"loc\\":{\\"lon\\":0,\\"lat\\":0},\\"tags\\":{},\\"type\\":\\"node\\"}",
]
`;

exports[`stringifying simple stringify 3`] = `
Array [
  Object {
    "attributes": Object {},
    "id": "n1",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {},
    "type": "node",
  },
]
`;
